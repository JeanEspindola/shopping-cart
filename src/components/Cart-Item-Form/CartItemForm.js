/*
* Cart Item Form Presentation Component.
*
* Renders the modal form to add a new product.
* There are not so many validations, product and comments are string fields,
* while price and tax are number. But I'm not validating the number of digits here.
* I assume the products have 2 decimal digits and taxes as only 7 or 19.
*
*/
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Modal, Button, Form, FormGroup, FormControl, Col, ControlLabel, InputGroup } from 'react-bootstrap';

class CartItemForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      product: '',
      comments: '',
      price: '',
      tax: '',
    };

    this.onChange = this.onChange.bind(this);
    this.saveProduct = this.saveProduct.bind(this);
    this.cancel = this.cancel.bind(this);
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  cleanUpState() {
    this.setState({
      product: '',
      comments: '',
      price: '',
      tax: '',
    });
  }

  saveProduct() {
    const productObject = {
      product: this.state.product,
      comments: this.state.comments,
      price: parseInt(this.state.price, 10),
      tax: parseInt(this.state.tax, 10),
    };

    this.props.onSaveItem(productObject);
    this.cleanUpState();
    this.props.onClose();
  }

  cancel() {
    this.cleanUpState();
    this.props.onClose();
  }

  render() {
    return (
      <Modal show={this.props.show}>
        <Modal.Header>
          <Modal.Title>{this.props.title}</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form horizontal>
            <FormGroup>
              <Col componentClass={ControlLabel} sm={2}>
                Product
              </Col>
              <Col sm={4}>
                <FormControl
                  type="text"
                  placeholder="Add your product"
                  value={this.state.product}
                  name="product"
                  onChange={this.onChange}
                />
              </Col>
            </FormGroup>
            <FormGroup>
              <Col componentClass={ControlLabel} sm={2}>
                Comments
              </Col>
              <Col sm={10}>
                <FormControl
                  type="text"
                  placeholder="Add comments for the product"
                  value={this.state.comments}
                  name="comments"
                  onChange={this.onChange}
                />
              </Col>
            </FormGroup>
            <FormGroup>
              <Col componentClass={ControlLabel} sm={2}>
                Price
              </Col>
              <Col sm={4}>
                <InputGroup>
                  <FormControl
                    type="number"
                    placeholder="Price"
                    value={this.state.price}
                    name="price"
                    onChange={this.onChange}
                  />
                  <InputGroup.Addon>$</InputGroup.Addon>
                </InputGroup>
              </Col>
            </FormGroup>
            <FormGroup>
              <Col componentClass={ControlLabel} sm={2}>
                Tax
              </Col>
              <Col sm={4}>
                <InputGroup>
                  <FormControl
                    type="number"
                    placeholder="Tax"
                    value={this.state.tax}
                    name="tax"
                    onChange={this.onChange}
                  />
                  <InputGroup.Addon>(7 or 19)%</InputGroup.Addon>
                </InputGroup>
              </Col>
            </FormGroup>
          </Form>
        </Modal.Body>
        <Modal.Footer>
          <Button bsStyle="primary" onClick={this.saveProduct}>Save</Button>
          <Button onClick={this.cancel}>Cancel</Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

CartItemForm.propTypes = {
  onClose: PropTypes.func.isRequired,
  show: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
  onSaveItem: PropTypes.func.isRequired,
};

export default CartItemForm;
