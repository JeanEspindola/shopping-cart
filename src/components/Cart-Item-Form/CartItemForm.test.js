import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import CartItemForm from './CartItemForm';

describe('CartItem', () => {
  const mockOnClose = jest.fn;
  const mockOnSaveItem = jest.fn;

  let Component;

  beforeEach(() => {
    Component = shallow(<CartItemForm
      show
      title="Add product"
      onClose={mockOnClose}
      onSaveItem={mockOnSaveItem}
    />);
  });

  it('renders CartItemForm without crashing', () => {
    expect(Component.length).toBeTruthy();
    expect(shallowToJson(Component)).toMatchSnapshot();
  });
});
