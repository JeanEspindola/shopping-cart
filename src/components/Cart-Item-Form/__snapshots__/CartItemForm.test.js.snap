// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem renders CartItemForm without crashing 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={false}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Add product
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <Form
      bsClass="form"
      componentClass="form"
      horizontal={true}
      inline={false}
    >
      <FormGroup
        bsClass="form-group"
      >
        <Col
          bsClass="col"
          componentClass={[Function]}
          sm={2}
        >
          Product
        </Col>
        <Col
          bsClass="col"
          componentClass="div"
          sm={4}
        >
          <FormControl
            bsClass="form-control"
            componentClass="input"
            name="product"
            onChange={[Function]}
            placeholder="Add your product"
            type="text"
            value=""
          />
        </Col>
      </FormGroup>
      <FormGroup
        bsClass="form-group"
      >
        <Col
          bsClass="col"
          componentClass={[Function]}
          sm={2}
        >
          Comments
        </Col>
        <Col
          bsClass="col"
          componentClass="div"
          sm={10}
        >
          <FormControl
            bsClass="form-control"
            componentClass="input"
            name="comments"
            onChange={[Function]}
            placeholder="Add comments for the product"
            type="text"
            value=""
          />
        </Col>
      </FormGroup>
      <FormGroup
        bsClass="form-group"
      >
        <Col
          bsClass="col"
          componentClass={[Function]}
          sm={2}
        >
          Price
        </Col>
        <Col
          bsClass="col"
          componentClass="div"
          sm={4}
        >
          <InputGroup
            bsClass="input-group"
          >
            <FormControl
              bsClass="form-control"
              componentClass="input"
              name="price"
              onChange={[Function]}
              placeholder="Price"
              type="number"
              value=""
            />
            <InputGroupAddon
              bsClass="input-group-addon"
            >
              $
            </InputGroupAddon>
          </InputGroup>
        </Col>
      </FormGroup>
      <FormGroup
        bsClass="form-group"
      >
        <Col
          bsClass="col"
          componentClass={[Function]}
          sm={2}
        >
          Tax
        </Col>
        <Col
          bsClass="col"
          componentClass="div"
          sm={4}
        >
          <InputGroup
            bsClass="input-group"
          >
            <FormControl
              bsClass="form-control"
              componentClass="input"
              name="tax"
              onChange={[Function]}
              placeholder="Tax"
              type="number"
              value=""
            />
            <InputGroupAddon
              bsClass="input-group-addon"
            >
              (7 or 19)%
            </InputGroupAddon>
          </InputGroup>
        </Col>
      </FormGroup>
    </Form>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
    >
      Save
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      onClick={[Function]}
    >
      Cancel
    </Button>
  </ModalFooter>
</Modal>
`;
